File: mpi/solutions/mpi_isend/mpi_isend.f90
PROGRAM mpi_isend_example
   USE mpi_f08
   IMPLICIT NONE

   INTEGER :: rank, size, ierr
   INTEGER :: tag, destination, count
   INTEGER :: buffer ! value to send
   INTEGER :: input_unit, io_status
   TYPE(MPI_Status) :: status
   TYPE(MPI_Request) :: request

   tag = 1234
   destination = 2 ! destination process
   count = 1 ! number of elements in buffer

! Turn it into an MPI program (initialise MPI)
   CALL MPI_Init(ierr)
   
   CALL MPI_Comm_size(MPI_COMM_WORLD, size, ierr) ! number of processes
   CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr) ! rank of current process

   IF (destination >= size) THEN
      CALL MPI_Abort(MPI_COMM_WORLD, 1, ierr) ! destination process must be under the number of processes created, otherwise abort
   END IF

   IF (rank == 0) THEN
      ! Open input file and read value
      OPEN(UNIT=10, FILE='input.txt', STATUS='OLD', ACTION='READ', IOSTAT=io_status)
      IF (io_status /= 0) THEN
         WRITE(*,*) 'Warning: Could not open input.txt, using default value'
         buffer = 42  ! Default value
      ELSE
         WRITE(*,'(A,I0,A)') 'Enter a value to send to processor ', destination, ':'
         READ(10, *, IOSTAT=io_status) buffer
         IF (io_status /= 0) THEN
            WRITE(*,*) 'Failed to read from input.txt, using default value'
            buffer = 42  ! Default value
         ELSE
            WRITE(*,'(A,I0)') 'Read value: ', buffer
         END IF
         CLOSE(10)
      END IF
      CALL MPI_Isend(buffer, count, MPI_INTEGER, destination, tag, MPI_COMM_WORLD, request, ierr) ! non blocking send to destination process
   END IF

   IF (rank == destination) THEN
      CALL MPI_Irecv(buffer, count, MPI_INTEGER, 0, tag, MPI_COMM_WORLD, request, ierr) ! destination process receives
   END IF

   CALL MPI_Wait(request, status, ierr) ! blocks and waits for destination process to receive data

   IF (rank == 0) THEN
      WRITE(*,'(A,I0,A,I0)') 'processor ', rank, ' sent ', buffer
   END IF
   IF (rank == destination) THEN
      WRITE(*,'(A,I0,A,I0)') 'processor ', rank, ' got ', buffer
   END IF

   CALL MPI_Finalize(ierr)

END PROGRAM mpi_isend_example

File: mpi/solutions/mpi_isend/mpi_isend.c
#include <stdio.h>
#include "mpi.h"


////////////
//MPI_Isend
////////////
//
// int MPI_Isend(const void *buf, int count, MPI_Datatype datatype, int dest, int tag,
//              MPI_Comm comm, MPI_Request *request)
//
// This example uses MPI_Isend to do a non-blocking send of information from the root process to a destination process.
// The destination process is set as a variable in the code and must be less than the number of processes started.
//
// example usage:
//		compile: mpicc -o mpi_isend mpi_isend.c
//		run: mpirun -n 4 mpi_isend
//
int main(argc, argv)
int argc;
char **argv;
{
    int rank, size;
    int tag, destination, count;
    int buffer; //value to send
    FILE *input_file = NULL;

    tag = 1234;
    destination = 2; //destination process
    count = 1; //number of elements in buffer    

    MPI_Status status;
    MPI_Request request = MPI_REQUEST_NULL;

    MPI_Init(&argc, &argv);

    MPI_Comm_size(MPI_COMM_WORLD, &size); //number of processes
    MPI_Comm_rank(MPI_COMM_WORLD, &rank); //rank of current process

    if (destination >= size) {
        MPI_Abort(MPI_COMM_WORLD, 1); // destination process must be under the number of processes created, otherwise abort
    }

    // Open input file once at the beginning (only for rank 0)
    if (rank == 0) {
        input_file = fopen("input.txt", "r");
        if (!input_file) {
            printf("Warning: Could not open input.txt, will use default value\n");
            buffer = 42; // Default value
        } else {
            printf("Enter a value to send to processor %d:\n", destination);
            if (fscanf(input_file, "%d", &buffer) == 1) {
                printf("Read value: %d\n", buffer);
            } else {
                printf("Failed to read from input.txt, using default value\n");
                buffer = 42; // Default value
            }
            fclose(input_file);
        }
        MPI_Isend(&buffer, count, MPI_INT, destination, tag, MPI_COMM_WORLD, &request); //non blocking send to destination process
        // MPI_Send(&buffer, count, MPI_INT, destination, tag, MPI_COMM_WORLD); //blocking send to destination process
    }

    if (rank == destination) {
        MPI_Irecv(&buffer, count, MPI_INT, 0, tag, MPI_COMM_WORLD, &request); //destination process receives
        // MPI_Recv(&buffer, count, MPI_INT, 0, tag, MPI_COMM_WORLD, &status); //destination process receives
    }

    MPI_Wait(&request, &status); //blocks and waits for destination process to receive data

    if (rank == 0) {
        printf("processor %d sent %d\n", rank, buffer);
    }
    if (rank == destination) {
        printf("processor %d got %d\n", rank, buffer);
    }

    MPI_Finalize();

    return 0;
}